torch/_export/graph_module.py

```diff
@@ -224,59 +194,88 @@ def op_to_str(op):
         return (reduce_graph_module, (pickled_state,))
 
 
-def attach_export_graph_metadata(gm: fx.GraphModule, meta: ExportMetadata) -> None:
-    gm.meta[EXPORT_METADATA] = meta
-    gm.__class__ = type(type(gm).__name__, (ExportGraphModuleMixin, type(gm)), {})
+class ExportedProgram:
+    def __init__(
+        self,
+        root: Union[torch.nn.Module, Dict[str, Any]],
+        graph: torch.fx.Graph,
+        graph_signature: ExportGraphSignature,
+        call_spec: CallSpec,
+        state_dict: Dict[str, Any],
+    ):
+        # Remove codegen related things from the graph. It should just be a flat graph.
+        graph._codegen = torch.fx.graph.CodeGen()
+        gm = torch.fx.GraphModule(root, graph, class_name="ExportGraphModule")
+        gm.__class__ = type(type(gm).__name__, (ExportGraphModuleMixin, type(gm)), {})
+        self.graph_module: ExportGraphModule = gm
+
+        self.graph_signature: ExportGraphSignature = graph_signature
+        self.call_spec: CallSpec = call_spec
+        self.state_dict: Dict[str, Any] = state_dict
+        self.symbol_to_range: Dict[str, Tuple[int, int]] = {}
+        self._input_shape_constraints: Dict[str, List[Tuple[int, ConstraintExpr, ConstraintExpr]]] = {}
+        self._input_name_to_example_inputs: Dict[str, Any] = {}
+
+    def __call__(self, *args: Any) -> Any:
+        if self.call_spec.in_spec is not None:
+            try:
+                args = fx_pytree.tree_flatten_spec(args, self.call_spec.in_spec)  # type: ignore[assignment]
+            except Exception as e:
+                raise error.InternalError("The in_spec is not correctly maintained.") from e
```

Comment:
I saw this error message before and it's kinda confusing from user perspective, consider something similar to the following:
```
Trying to flatten user inputs with exported input tree spec:
{self.in_spec}
but actually got inputs with tree spec of:
{spec_from_flattening_args}
```